File: Mosh Hamadeni Course/Linked Lists/linked_list.py
class LinkedList():

  class Node():
    next_node = None
    def __init__(self, value):
      self.value = value

  first = None
  last = None
  length = 0


  def add_last(self, item):
    self.node = self.Node(item)
    self.length += 1

    if self.is_empty():
      self.first = self.last = self.node
    else:
      self.last.next_node = self.node
      self.last = self.node

  def add_first(self, item):
    self.node = self.Node(item)
    self.length += 1

    if self.is_empty():
      self.first = self.last = self.node
    else:
      self.node.next_node = self.first
      self.first = self.node
  
  def is_empty(self):
    return self.first == None

  def index_of(self, item):
    self.index = 0
    self.current = self.first
    while self.current != None:
      if self.current.value == item:
        return self.index
      else:
        self.current = self.current.next_node
        self.index += 1
    else: return -1
    
  def contains(self, item):
    return self.index_of(item) != -1

  def remove_first(self):
    if self.is_empty():
      raise Exception('Linked list is Empty')
    elif self.first == self.last:
      self.first = self.last = None
      self.length -= 1
    else:
      self.second = self.first.next_node
      self.first.next_node = None
      self.first = self.second
      self.length -= 1






    if self.first and self.first.next_node:
      self.second = self.first.next_node
      self.first.next_node = None
      self.first = self.second
      self.length -= 1




linked_list = LinkedList()
linked_list.add_last(10)
linked_list.add_last(20)
# linked_list.add_last(30)
# linked_list.add_first(50)
linked_list.remove_first()

print(linked_list.first.value)
# print(linked_list.first.next_node.value)
# print(linked_list.last.value)
# print(linked_list.last.next_node)
print(linked_list.length)
# print(linked_list.index_of(100))
# print(linked_list.contains(100))

File: Mosh Hamadeni Course/JavaScript/linkedList.js
class LinkedList {
  Node = class {
    value;
    next = null;
    constructor(value) {
      this.value = value;
    }
  };

  first;
  last;
  length = 0;

  // addLast
  addLast(item) {
    const node = new this.Node(item);
    // console.log(node);

    if (this.isEmpty()) {
      this.first = this.last = node;
    } else {
      this.last.next = node;
      this.last = node;
    }
    // console.log(this.first);
    // console.log(this.last);
    this.length++;
  }

  // addFirst
  addFirst(item) {
    const node = new this.Node(item);

    if (this.isEmpty()) {
      this.first = this.last = node;
    } else {
      node.next = this.first;
      this.first = node;
    }
    console.log(this.first);
    console.log(this.last);
    this.length++;
  }

  // assAt
  addAt(item, index) {
    if (index > this.length) throw new Error('Index out of range');

    const node = new this.Node(item);
    let before;
    let current = this.first;
    for (let i = 0; i < this.length; i++) {
      if (i === index - 1) {
        before = current;
      }
      if (i === index) {
        before.next = node;
        node.next = current;
        break;
      }
      current = current.next;
    }
    this.length++;
    console.log(this.first);
    // console.log(this.length);
  }

  // deleteFirst

  deleteFirst() {
    if (this.length === 0) throw new Error('Linked list is empty');

    const second = this.first.next;
    this.first.next = null;
    this.first = second;
    this.length--;
    console.log(this.first);
  }

  // deleteLast
  deleteLast() {
    let current = this.first;

    for (let i = 2; i < this.length; i++) {
      current = current.next;

      if (i === this.length - 1) {
        this.last = current;
        this.last.next = null;
      }
    }
    this.length--;
    console.log(this.first);
    console.log(this.last);
  }

  // contains
  // indexOf

  isEmpty() {
    return this.first === undefined;
  }
}

const list = new LinkedList();
list.addLast(10);
list.addLast(20);
list.addLast(30);
list.addFirst(-10);
// list.addAt(100, 1);
// list.deleteFirst();
list.deleteLast();
